openwisp_users/api/views.py

```diff
@@ -20,4 +39,62 @@ def post(self, request, *args, **kwargs):
         return super().post(request, *args, **kwargs)
 
 
+class ProtectedAPIMixin(object):
+    authentication_classes = [BearerAuthentication, SessionAuthentication]
+    permission_classes = [
+        IsAuthenticated,
+        DjangoModelPermissions,
+    ]
+
+
+class BaseOrganizationView(ProtectedAPIMixin):
+    serializer_class = OrganizationSerializer
+
+    def get_queryset(self):
+        user = self.request.user
+        if user.is_superuser:
+            return Organization.objects.order_by('-created')
+        return Organization.objects.none()
+
+
+class OrganizationListCreateView(BaseOrganizationView, ListCreateAPIView):
+    pagination_class = ListViewPagination
+
+
+class OrganizationDetailView(BaseOrganizationView, RetrieveUpdateDestroyAPIView):
+    pass
+
+
+class BaseUserView(ProtectedAPIMixin):
+    serializer_class = UserSerializer
+
+    def get_queryset(self):
+        user = self.request.user
+        if user.is_superuser:
+            return User.objects.all()
+
+        if not user.is_superuser:
+            org_users = OrganizationUser.objects.filter(user=user).select_related(
+                'organization'
+            )
+            qs = User.objects.none()
+            for org_user in org_users:
+                if org_user.is_admin:
+                    qs = qs | org_user.organization.users.all().distinct()
```

Comment:
Do we need a `distinct` here?

Reply:
I have implemented this logic from here https://github.com/openwisp/openwisp-users/blob/master/openwisp_users/multitenancy.py#L99-L119

Reply:
If the code is trying to do something so similar at 2 places, why is it not shared? 
i.e creating a mixin to share the logic appropriately! :smile:

Reply:
Sure, @atb00ker , I will  try to move this logic to a mixin. ASAP. :smile: